### PR TITLE
mongodb: 3.4.10 -> 4.0.12

### PR DESCRIPTION
**Pull request by jarjee**
###### Motivation for this change
Bringing the mongodb server up to the latest stable release. Also had some odd issues with the JS interpreter when using "load()", which seems to be fixed in this version.

###### Things done
- [X] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `build-use-sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [X] NixOS
   - [x] macOS
   - [ ] other Linux distributions
- [x] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [X] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [X] Tested execution of all binary files (usually in `./result/bin/`)
- [X] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

### CODE REVIEW

**Comment by jarjee** [comment]
@GrahamcOfBorg build mongodb

**Comment by jarjee** [comment]
cc @offlinehacker @bluescreen303 @wkennington @cstrahan

**Comment by obadz** [comment]
@GrahamcOfBorg build mongodb

**Review by obadz — CHANGES_REQUESTED** [review]
Needs darwin fix

**Review by obadz — CHANGES_REQUESTED** [review]
Also needs aarch64 to be marked broken (or fixed)

**Comment by jarjee** [comment]
I don't have access to a mac to test with, but adding `libtools` should do it.

**Comment by jarjee** [comment]
@GrahamcOfBorg build mongodb

**Comment by obadz** [comment]
@GrahamcOfBorg build mongodb

**Comment by xeji** [comment]
@GrahamcOfBorg  build mongodb

**Comment by cyounkins** [comment]
Hello! @jarjee could you rebase this, or may I make a new PR with your changes?

Also, Mongo is no longer supported on 32-bit systems - "Changed in version 3.4: MongoDB no longer supports 32-bit x86 platforms." so I think `meta.platforms` should be set to `stdenv.lib.subtractLists platforms.doubles.i686 platforms.doubles.unix;` (Reference #52463)

**Comment by cyounkins** [comment]
It appears the 32-bit issue remains in your most recent push - We need to disable building on 32-bit systems as it isn't supported and in my testing has failed. This is the error that occurs: https://jira.mongodb.org/browse/SERVER-22810

Could you please set `meta.platforms` to `stdenv.lib.subtractLists platforms.doubles.i686 platforms.doubles.unix;` ? Thanks!

**Comment by jarjee** [comment]
@cyounkins Sorry for the wait, I've made the change you requested.

**Comment by ryantm** [comment]
@GrahamcOfBorg build mongodb

File: lib/licenses.nix [hunk 1]
```diff
@@ -624,6 +624,12 @@ lib.mapAttrs (n: v: v // { shortName = n; }) rec {
     url = http://metadata.ftp-master.debian.org/changelogs/main/d/debianutils/debianutils_4.8.1_copyright;
   };
 
+  sspl = {
+    shortName = "SSPL";
+    fullName = "Server Side Public License";
+    url = https://www.mongodb.com/licensing/server-side-public-license;
+  };
```
Comment (schmittlauch) [inline lib/licenses.nix]: This lacks a `free = false;`, as the SSPL is not recognized by the OSI and also has been [dropped from Fedora for that reason.](https://www.geekwire.com/2019/mongodbs-licensing-changes-led-red-hat-drop-database-latest-version-server-os/)

**Comment by dpausp** [comment]
I'm not an expert for MongoDB, but the docs say that upgrades to 4.0 can only be done from an existing 3.6 installation. 3.4 should be left as default and 3.6 and 4.0 should be provided as additional versions to make upgrades possible, I think.

**Comment by aanderse** [comment]
@jarjee and/or @cyounkins are you able to address the above mentioned issue(s) and continue with this PR?

**Comment by aanderse** [comment]
@jarjee great, thanks! Let us know if you have any issues. Looking forward to getting this merged!

**Comment by jarjee** [comment]
I've just noticed #35298, so I'll also have a look and see if we need any of those patches.

Since we're also adding new packages, I'm thinking we'll also need to look at updating the tests so we test the different mongodb versions as well.

**Comment by aanderse** [comment]
@jarjee Yeah that sounds good. If you get stuck at any point please do ping this thread for help as I believe there are a number of people here with motivation to see this through. Thanks again for your work on this.

**Comment by aanderse** [comment]
Thanks @jarjee. What is the status on this now? Are you waiting on anyone/anything? Freeze for `19.09` has been announced for September 7th.

**Comment by jarjee** [comment]
Hi @aanderse Only issue that I currently have now is that this doesn't build on OSX. I just got access to a mac with limited access, so I've been trying to debug the build issues there.

**Comment by aanderse** [comment]
@GrahamcOfBorg build mongodb-3_4 mongodb-3_6 mongodb-4_0

**Comment by aanderse** [comment]
@jarjee you are awesome! Does this work on `darwin` now, do you need some assistance, etc...?

Thank you for your amazing work!

**Comment by jarjee** [comment]
@aanderse I'm still fiddling with 4.0, but 3.4 & 3.6 seem to be building now

File: pkgs/servers/nosql/mongodb/mongodb.nix [hunk 2]
```diff
@@ -97,9 +105,10 @@ in stdenv.mkDerivation rec {
   meta = {
     description = "A scalable, high-performance, open source NoSQL database";
     homepage = http://www.mongodb.org;
-    license = licenses.agpl3;
+    license = licenses.sspl;
```
Comment (alyssais) [inline pkgs/servers/nosql/mongodb/mongodb.nix]: Won’t this mean that AGPL versions of MongoDB are marked as non-free?

Comment (jarjee) [reply pkgs/servers/nosql/mongodb/mongodb.nix]: Good question, according to the [MongoDB README](https://github.com/mongodb/mongo/blob/b73a7d6ab9d38382a6c2f1fd0b5c8369d54d8b95/README#L89):

```
LICENSE

  MongoDB is free and the source is available. Versions released prior to
  October 16, 2018 are published under the AGPL. All versions released after
  October 16, 2018, including patch fixes for prior versions, are published
  under the Server Side Public License (SSPL) v1. See individual files for
  details.
```

Versions [3.4.22](https://docs.mongodb.com/manual/release-notes/3.4/#aug-6-2019), [3.6.13](https://docs.mongodb.com/manual/release-notes/3.6/#jun-10-2019) and [4.0.11](https://docs.mongodb.com/manual/release-notes/4.0/#jul-26-2019) were all released after October 16, 2018; so they all have to be marked as SSPL.

**Comment by jarjee** [comment]
Current state of this PR: I have 4.0, 3.6 and 3.4 building on OSX now :tada:. They don't seem to build in sandbox mode, I'm not sure if this is a problem / how to resolve this.

The other things that I'd like to do:
- [ ] Update v4_0 to use 4.0.12
- [ ] Clean up the nix functions, since there are variables that I'm pretty sure aren't actually being used in the builds

File: nixos/tests/mongodb.nix [hunk 3]
```diff
@@ -12,29 +12,76 @@ in {
...
+    };
+
+
+  };
+
   testScript = ''
-    startAll;
-    $one->waitForUnit("mongodb.service");
-    $one->succeed("mongo -u nixtest -p nixtest nixtest ${testQuery}") =~ /hello/ or die;
+         $v3_4->start;
```
Comment (offlinehacker) [inline nixos/tests/mongodb.nix]: I think it would be better to have generic testScript, that's version independent.

**Review by offlinehacker — COMMENTED** [review]
Why are you using multiple nodes in what is basically a single test with different parameters? I would much rather prefer running a single node for single mongodb version and split into multiple tests.

**Comment by aanderse** [comment]
From my perspective we're almost out of time on this. The freeze is coming and this is *very* important to merge. Thank you **so** much @jarjee for your work on this! I'm planning on merging this in about a weeks time. Hopefully everything at that point will be acceptable to all parties involved. @offlinehacker if the tests are modified to your specification will you be satisfied?

File: pkgs/servers/nosql/mongodb/mongodb.nix (hunk 2 above)
Comment (alyssais) [reply pkgs/servers/nosql/mongodb/mongodb.nix]: In that case, I’d be inclined to either not offer patch versions after that date, or to at least continue to offer the last free version.

**Comment by jarjee** [comment]
So the most recent openssl update seems to have broken 3_4, but I've now moved all the tests to use a single machine in 211f84c, based off the feedback from @offlinehacker.

**Comment by jarjee** [comment]
I've tried using the old version of openssl with 3_4, but it still doesn't seem to build. Looks like somehow the newer version of openssl is still getting included, even with passing down the old version explicitly.

**Comment by jarjee** [comment]
Is it worth adding AGPL variants? Essentially what  @alyssais suggested? I was inclined to not do that since it means that the packages will no longer update, but SSPL being non-free is causing issues. Especially since hydra doesn't run unfree tests AFAIK.

**Comment by aanderse** [comment]
@jarjee if only `3.4` is broken then maybe it is worth adding AGPL variant for that version alone. We're so close here...

**Comment by alyssais** [comment]
> @jarjee if only `3.4` is broken then maybe it is worth adding AGPL variant for that version alone. We're so close here...

I think 4.1, being the last release series to be available as AGPL,
should have an AGPL variant available in Nixpkgs.

**Comment by tricktron** [comment]
I fixed the ssl error of `v.3.4` in #81027. How could I have added the commits to this pr instead of creating a new one?

**Comment by jarjee** [comment]
Hi @tricktron sorry for the wait, thanks for making the fix. I can cherry pick your PRs across if that's ok with you.

If you wanted to add these commits to this PR instead of making a new one, you'd need to add them to my forked version of nixpkgs (adding the commits to [`jarjee/nixpkgs:update/mongodb`](https://github.com/jarjee/nixpkgs/tree/update/mongodb))

**Comment by tricktron** [comment]
> Hi @tricktron sorry for the wait, thanks for making the fix. I can cherry pick your PRs across if that's ok with you.

Yeah, that is fine!

> If you wanted to add these commits to this PR instead of making a new one, you'd need to add them to my forked version of nixpkgs (adding the commits to [`jarjee/nixpkgs:update/mongodb`](https://github.com/jarjee/nixpkgs/tree/update/mongodb))

Makes sense, thanks for the info.

**Comment by uri-canva** [comment]
Tested this on macOS and it works.

**Comment by dasJ** [comment]
This is proably more important than expected (also for backport to 20.03) because the mongodb package does not build with:
```
in job ‘nixos.tests.mongodb.x86_64-linux’:
Package ‘openssl-1.0.2u’ in /nix/store/76gxrrf0jmapcvw25prlqi6m89aa8q3p-source/pkgs/development/libraries/openssl/default.nix:130 is marked as insecure, refusing to evaluate.


Known issues:
 - Support for OpenSSL 1.0.2 ended with 2019.

You can install it anyway by whitelisting this package, using the
following methods:
```
The MongoDB package on both master and 20.03 does not build with the newer OpenSSL 1.1
As NixOS should probably keep support for MongoDB, this needs to be both merged and backported.

**Review by Ma27 — APPROVED** [review]
LGTM. As @dasJ already mentioned, this should go into 20.03, otherwise we'll only ship unsecure versions of mongodb which isn't a good idea IMHO.

I just tested `mongodb` with the aarch64 commnunity builder and it appears to work fine, so unmarked it as broken.

Are there any more things that would keep us from merging this PR?

@GrahamcOfBorg test mongodb

**Comment by jul1u5** [comment]
Ping. This would be very useful. Can we merge this?

**Comment by aanderse** [comment]
I would love that. I lost track where we're at with this, though. Are you running with this now @Ma27?

**Comment by Ma27** [comment]
> Are you running with this now

In `#nixos-dev` everybody seemed to be fine with this. Will merge (and backport mongodb-3.4.24) tomorrow :)

**Comment by Ma27** [comment]
I rewrote the test in python, rebased to master and cleaned up the history. As soon as the mongodb builds (and test) I'm running locally are finished, I'll push the changeset and merge :)

**Comment by Ma27** [comment]
@GrahamcOfBorg test mongodb

**Comment by Ma27** [comment]
@GrahamcOfBorg build mongodb-4_0 mongodb-3_4 mongodb-3_6

**Comment by Ma27** [comment]
@jarjee @worldofpeace @offlinehacker @aanderse can you please take a look at the license (sspl)?

According to MongoDB's FAQ it's free, based on GPLv3, but not (yet?) approved by the OSI (https://www.mongodb.com/licensing/server-side-public-license/faq).

From my understanding it should be therefore possible to mark `licenses.ssdp` as free, however I'm definetely not a software licensing expert or a lawyer. Or do we only mark OSI-approved licenses as free? (Didn't find any reference about this while skimming through the manual).

The main reason I'm asking is because it takes quite some time and resources to compile `mongodb` and the unfree licenses would mean that everyone has to compile it locally.

I guess one of the most important paragraphs from the FAQ is the following:

> Although the SSPL is not OSI approved, it maintains all of the same freedoms the community has always had with MongoDB under AGPL. Users are free to review, modify, and distribute the software or redistribute modifications to the software. However, the Open Source Initiative (OSI) has its own process for approving what it considers to be an open source license, and the SSPL has not received OSI approval. MongoDB software licensed under the SSPL is not considered open source by the OSI.

**Comment by Ma27** [comment]
Merging for now. As discussed in [`#nixos-dev`](https://logs.nix.samueldr.com/nixos-dev/2020-03-26#3227422;), this is at least an improvement over the current state. We need to find a solution for the licensing issue soon though.

**Comment by adrian-gierakowski** [comment]
Thanks for this! Any idea if it would take much work to also add mongo 4.2?

**Comment by Ma27** [comment]
Without having it tested: I don't think so, the packages are already splitted, so it should be fairly easy to add another one :)

For the record: I didn't forget the backport, I'll file a PR for this tomorrow hoping that I get a bit more feedback on #83433 first :)

**Pull request merged**